### PR TITLE
Compare View for big files with Line Spacing freezes eclipse

### PR DESCRIPTION
**Pull request by DenisUngemach**
For big files with many differences, the compare view freezed eclipse, if the line spacing was set. The reason were the necessary resources to calculate the spacing in the method getHeightBetweenLines. We have to save the heights in a list and use them to increase the performance.

Fixes https://github.com/eclipse-platform/eclipse.platform.ui/issues/2143

### CODE REVIEW

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java [hunk 1]
```diff
@@ -474,6 +475,12 @@ private static RGB interpolate(RGB fg, RGB bg, double scale) {
 	private IUndoableOperation copyUndoable = null;
 	private IOperationHistoryListener operationHistoryListener;
 
+	// https://github.com/eclipse-platform/eclipse.platform.ui/issues/2143
+	// calling w.getLineHeight(line) + w.getLineSpacing() +
+	// w.getLineVerticalIndent(line); again and again for the same line is too slow
+	// we therefore introduce a line height cache by viewer
```
Comment (iloveeclipse) [inline team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:481]: Should it be a field, or can it be local variable?

Users can edit & save files in compare editor & so may add/remove lines.
Would this map **as field* mean, the line height will be wrong for the same compare editor after changes?

Comment (DenisUngemach) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:481]: It must be a field, else the recalculation as local variable need too much time. 

Because of textChanges, there is a listener in the method: getHeightBetweenLines which reacts on text changes and clears this field completely.

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java [hunk 2]
```diff
@@ -670,6 +673,35 @@ private void internalSetSelection(TreeItem ti, boolean fireOpen) {
 		}
 	}
 
+	private void expandCollapseAction(DoubleClickEvent event) {
```
Comment (iloveeclipse) [inline team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java:676]: What is the purpose of this change?

Comment (DenisUngemach) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java]: I have deleted that sorry for that.
I hate github....

Comment (DenisUngemach) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java]: This is the next issue: https://github.com/eclipse-platform/eclipse.platform.ui/issues/1838 
There i create a new pull request.

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java [hunk 3]
```diff
@@ -4204,11 +4211,31 @@ private void updateLines(IDocument d) {
 	 * @param line
 	 * @return
 	 */
-	private int getHeightBetweenLines(MergeSourceViewer tp, int fromLine, int toLine) {
+	int getHeightBetweenLines(MergeSourceViewer tp, int fromLine, int toLine) {
```
Comment (BeckerWdf) [inline team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:4214]:
```suggestion
	private int getHeightBetweenLines(MergeSourceViewer tp, int fromLine, int toLine) {
```

**Review by tobiasmelcher — CHANGES_REQUESTED** [review]
I think the newly introduced cache lineHeightsByViewer needs to be cleared in method handlePropertyChangeEvent for the scenario when the line spacing settings are changed in the preferences.

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java (hunk 3 above)
Comment (DenisUngemach) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:4214]: Sure.

**Comment by DenisUngemach** [comment]
> I think the newly introduced cache lineHeightsByViewer needs to be cleared in method handlePropertyChangeEvent for the scenario when the line spacing settings are changed in the preferences.

This recommendation works: if i change the spacing in the Text Editor preferences, the line drawings will be updated properly.
But the call of handlePropertyChangeEvent doesn't change the spacing in the StyledTexts, , this happens in a completely different layer.

Which means this works, because the call handlePropertyChangeEvent in TextMergeViewer happens before StyledText.setLineSpacing(int) .

The actual change of the line spacing comes from AbstractTextEditor.PropertyChangeListener.propertyChange(PropertyChangeEvent)

In theory this could lead to line drawing bugs, but i wouldn't expect it.

Is this ok for everybody? 
It is also possible set the spacing and other line height properties directly in the TextMergeViewer.handlePropertyChangeEvent , but this is bad design, because then the StyledText-Properties will be modified twice.

It seems there is no StyledText-Listener where i can check whether the properties were changed and clear  lineHeightsByViewer  afterwards.

**Comment by iloveeclipse** [comment]
> Is this ok for everybody?

OK for me. May be a comment could be added to the current listener?

**Comment by DenisUngemach** [comment]
Ok, done.

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java (hunk 3 above)
Comment (BeckerWdf) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:4224]: "private" is still missing.

**Review by BeckerWdf — CHANGES_REQUESTED** [review]

**Review by BeckerWdf — APPROVED** [review]

**Review by tobiasmelcher — APPROVED** [review]
thanks a lot Denis. Looks fine from my side.
